server/cluster/cluster.go

```diff
@@ -22,24 +23,46 @@ type metaData struct {
 	clusterTopology *clusterpb.ClusterTopology
 }
 
+func (m metaData) GetCluster() *clusterpb.Cluster {
+	return m.cluster
+}
+
+func (m metaData) GetClusterTopology() *clusterpb.ClusterTopology {
+	return m.clusterTopology
+}
+
 type Cluster struct {
 	clusterID uint32
 
 	// RWMutex is used to protect following fields.
 	lock         sync.RWMutex
 	metaData     *metaData
-	shardsCache  map[uint32]*Shard  // shard_id -> shard
-	schemasCache map[string]*Schema // schema_name -> schema
-	nodesCache   map[string]*Node   // node_name -> node
+	shardsCache  map[uint32]*Shard         // shard_id -> shard
+	schemasCache map[string]*Schema        // schema_name -> schema
+	nodesCache   map[string]*Node          // node_name -> node
+	shardLockMap map[uint32]*ShardWithLock // shard_id -> shardLock
```

Comment:
I guess this should be encapsulated as a specific struct called `EntryLock`, so that unit tests will be easy to design for it. One `TODO` can be left here if you don't do this in the PR.